### ISSUE TITLE
One stolon-keeper fails to start

### ISSUE THREAD

**Issue by nh2** [sorintlab/stolon#359]
In a 3-machine synchronous cluster, one machine is now failing to start its `stolon-keeper`:

```
# sudo -u postgres /nix/store/ngy0lpk389vbn74yjhf2i5f6f192q1hv-stolon-0.6.0/bin/stolon-keeper --cluster-name test-stolon-cluster --store-backend=consul --uid node_1 --data-dir /var/lib/postgres/node_1 --pg-listen-address=10.0.0.1 --pg-bin-path=/nix/store/qq2fxv16cm5r6wshk5hbiwn284s16h1n-postgresql-9.6.3/bin --pg-su-username=postgres --pg-su-passwordfile=/etc/stolon/test-stolon-cluster.password --pg-repl-username=postgres --pg-repl-passwordfile=/etc/stolon/test-stolon-cluster.password

warning: superuser name and replication user name are the same. Different users are suggested.
[I] 2017-09-28T01:05:01Z keeper.go:1567: exclusive lock on data dir taken
[I] 2017-09-28T01:05:01Z keeper.go:408: keeper uid uid=node_1
[I] 2017-09-28T01:05:02Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-09-28T01:05:02Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-09-28T01:05:02Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-09-28T01:05:02Z keeper.go:839: our db boot UID is different than the cluster data one, waiting for it to be updated bootUUID=3daf23f4-2543-49ad-91e8-3466a07a0c20 clusterBootUUID=7e395ead-a25d-4612-ae09-828eae2b5194
[I] 2017-09-28T01:05:02Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-09-28T01:05:03Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-09-28T01:05:03Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-09-28T01:05:03Z keeper.go:839: our db boot UID is different than the cluster data one, waiting for it to be updated bootUUID=3daf23f4-2543-49ad-91e8-3466a07a0c20 clusterBootUUID=7e395ead-a25d-4612-ae09-828eae2b5194
[I] 2017-09-28T01:05:03Z postgresql.go:215: stopping database
...
```

There is no postgresql process.

What might be the issue here?

And how might I get that machine back into operating normally?

**Comment by nh2** [comment]
Looking at `strace -e execve` output, it doesn't seem to try and start postgres; the only `execve()` syscall that it repeats is:

```
[pid 25964] execve("/nix/store/qq2fxv16cm5r6wshk5hbiwn284s16h1n-postgresql-9.6.3/bin/pg_ctl", ["/nix/store/qq2fxv16cm5r6wshk5hbiwn284s16h1n-postgresql-9.6.3/bin/pg_ctl", "stop", "-w", "-D", "/var/lib/postgres/node_1/postgres", "-o", "-c unix_socket_directories=/tmp", "-m", "fast"], [/* 20 vars */]) = 0
```

**Comment by sgotti** [comment]
Looks like the keeper has no db assigned by the sentinel so it doesn't have start anything . Can you provide the output of  `stolonctl clusterdata` and the sentinel logs if there's something useful in them?

P.S. Probably such kind of questions will be better if discussed on the stolon mailing list or gitter channel.

**Comment by sgotti** [comment]
@nh2 ping

**Comment by nh2** [comment]
Tis one will take me a bit longer as I first have to upgrade to 0.7 to be able to use `stolonctl clusterdata`.

**Comment by nh2** [comment]
@sgotti Do I need `stolonctl clusterdata` or can I also paste the JSON value directly out of Consul?

Also, how does the UUID logic work?

I see that the UUID is created in https://github.com/sorintlab/stolon/blob/v0.6.0/cmd/keeper/keeper.go#L363.

Doesn't this run every time stolon-keeper starts up, and as such, creates a new random UUID every time? What ensures that this UUID is equal to what's in the cluster data?

**Comment by nh2** [comment]
I noticed that every time I restart stolon-keeper, the UUIDs "rotate"; across 3 restarts:

```
bootUUID=5bd79f80-f7ff-465e-be90-e0b25d0f13ac clusterBootUUID=04cd6f50-0b5d-46e9-b883-01de9609bddf
bootUUID=7672a2be-48ed-41e4-896f-4ce1a39664ee clusterBootUUID=5bd79f80-f7ff-465e-be90-e0b25d0f13ac
bootUUID=36c153ee-5532-4633-acf0-81c7a268aedd clusterBootUUID=7672a2be-48ed-41e4-896f-4ce1a39664ee
```

What's `bootUUID` in the last start becomes `clusterBootUUID` at the next boot.

So when the message says `waiting for it to be updated`, it seems that eventually happens.

**Comment by nh2** [comment]
By now a second machine has failed with the same problem, rendering the cluster down (2 of 3 down).

I also gathered some info from the node when starting `stolon-keeper` manually:

```
# sudo -u postgres /nix/store/ngy0lpk389vbn74yjhf2i5f6f192q1hv-stolon-0.6.0/bin/stolon-keeper --cluster-name test-stolon-cluster --store-backend=consul --uid node_1 --data-dir /var/lib/postgres/node_1 --pg-listen-address=10.0.0.1 --pg-bin-path=/nix/store/qq2fxv16cm5r6wshk5hbiwn284s16h1n-postgresql-9.6.3/bin --pg-su-username=postgres --pg-su-passwordfile=/etc/stolon/test-stolon-cluster.password --pg-repl-username=postgres --pg-repl-passwordfile=/etc/stolon/test-stolon-cluster.password
warning: superuser name and replication user name are the same. Different users are suggested.
[I] 2017-10-13T13:52:39Z keeper.go:1567: exclusive lock on data dir taken
[I] 2017-10-13T13:52:39Z keeper.go:408: keeper uid uid=node_1
[I] 2017-10-13T13:52:39Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-10-13T13:52:39Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T13:52:39Z keeper.go:839: our db boot UID is different than the cluster data one, waiting for it to be updated bootUUID=4452a0f6-779a-4df6-9e1e-82824791c796 clusterBootUUID=36c153ee-5532-4633-acf0-81c7a268aedd
[I] 2017-10-13T13:52:39Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-10-13T13:52:40Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-10-13T13:52:40Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T13:52:40Z keeper.go:839: our db boot UID is different than the cluster data one, waiting for it to be updated bootUUID=4452a0f6-779a-4df6-9e1e-82824791c796 clusterBootUUID=36c153ee-5532-4633-acf0-81c7a268aedd
[I] 2017-10-13T13:52:40Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-10-13T13:52:41Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-10-13T13:52:41Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T13:52:41Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[E] 2017-10-13T13:52:41Z keeper.go:1285: cannot move from master role to standby role
[E] 2017-10-13T13:52:42Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-10-13T13:52:42Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T13:52:42Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[E] 2017-10-13T13:52:42Z keeper.go:1285: cannot move from master role to standby role
^C
```

`cannot move from master role to standby role` suggests all down standbys think that they are master, and thus refuse to become standbys.

Interestingly, on the node that is actually the master, `recovery.conf` does not exist in the postgres data dir, only `recovery.done`

ON the two failed standbys, `recovery.conf` exists and is empty, and `recovery.done` also exists.

The master node is `node_3`, the two failed nodes are `node_1` and `node_2`.

Consider:

```
[root@node-1:~]# ls -l /var/lib/postgres/node_1/postgres/recovery*
-rw------- 1 postgres nogroup   0 Sep 21 21:34 /var/lib/postgres/node_1/postgres/recovery.conf
-rw------- 1 postgres nogroup 223 Sep 20 19:05 /var/lib/postgres/node_1/postgres/recovery.done

[root@node-1:~]# cat /var/lib/postgres/node_1/postgres/recovery.done 
primary_slot_name = 'stolon_a07b53fc'
recovery_target_timeline = 'latest'
standby_mode = 'on'
primary_conninfo = 'sslmode=disable user=postgres password=... host=10.0.0.1 port=5432 application_name=stolon_a07b53fc'
```

```
[root@node-2:~]# ls -l /var/lib/postgres/node_2/postgres/recovery*
-rw------- 1 postgres nogroup   0 Oct 11 20:30 /var/lib/postgres/node_2/postgres/recovery.conf
-rw------- 1 postgres nogroup 223 Oct 11 13:05 /var/lib/postgres/node_2/postgres/recovery.done

[root@node-2:~]# cat /var/lib/postgres/node_2/postgres/recovery.done 
standby_mode = 'on'
primary_conninfo = 'port=5432 application_name=stolon_e8a0366d sslmode=disable user=postgres password=... host=10.0.0.2'
primary_slot_name = 'stolon_e8a0366d'
recovery_target_timeline = 'latest'
```

```
[root@node-3:~]# ls -l /var/lib/postgres/node_3/postgres/recovery*
-rw------- 1 postgres nogroup 223 Sep 28 03:07 /var/lib/postgres/node_3/postgres/recovery.done

[root@node-3:~]# cat /var/lib/postgres/node_3/postgres/recovery.done 
standby_mode = 'on'
primary_conninfo = 'port=5432 application_name=stolon_e8a0366d sslmode=disable user=postgres password=... host=10.0.0.2'
primary_slot_name = 'stolon_e8a0366d'
recovery_target_timeline = 'latest'
```

How can it occur that the two standbys have empty `recovery.conf` files (thus stolon-keeper thinks these postgres instances are masters), but the cluster data shows them as `"role": "standby"`?

**Comment by nh2** [comment]
I've worked around the issue, bringing the standbys back to live, by moving `recovery.done` to `recovery.conf`:

```
[root@node-1:~]# systemctl stop stolon-keeper.service

[root@node-1:~]# mv /var/lib/postgres/node_1/postgres/recovery.done /var/lib/postgres/node_1/postgres/recovery.conf 

[root@node-1:~]# sudo -u postgres /nix/store/ngy0lpk389vbn74yjhf2i5f6f192q1hv-stolon-0.6.0/bin/stolon-keeper --cluster-name stolon-cluster --store-backend=consul --uid node_1 --data-dir /var/lib/postgres/node_1 --pg-listen-address=10.0.0.1 --pg-bin-path=/nix/store/qq2fxv16cm5r6wshk5hbiwn284s16h1n-postgresql-9.6.3/bin --pg-su-username=postgres --pg-su-passwordfile=/etc/stolon/stolon-cluster.password --pg-repl-username=postgres --pg-repl-passwordfile=/etc/stolon/stolon-cluster.password
warning: superuser name and replication user name are the same. Different users are suggested.
[I] 2017-10-13T14:37:53Z keeper.go:1567: exclusive lock on data dir taken
[I] 2017-10-13T14:37:53Z keeper.go:408: keeper uid uid=node_1
[I] 2017-10-13T14:37:53Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-10-13T14:37:53Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T14:37:53Z keeper.go:839: our db boot UID is different than the cluster data one, waiting for it to be updated bootUUID=f1ba4d9c-11f8-411f-a414-494afd404a89 clusterBootUUID=717c34f4-41c4-4506-9c5b-a041834d8bfb
[I] 2017-10-13T14:37:53Z postgresql.go:215: stopping database
pg_ctl: PID file "/var/lib/postgres/node_1/postgres/postmaster.pid" does not exist
Is server running?
[E] 2017-10-13T14:37:53Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-10-13T14:37:54Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[I] 2017-10-13T14:37:54Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[I] 2017-10-13T14:37:54Z keeper.go:1288: already standby
[I] 2017-10-13T14:37:54Z postgresql.go:199: starting database
waiting for server to start....[E] 2017-10-13T14:37:54Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
[E] 2017-10-13T14:37:55Z keeper.go:526: cannot get configured pg parameters error=dial tcp [::1]:5432: getsockopt: connection refused
LOG:  redirecting log output to logging collector process
HINT:  Future log output will appear in directory "pg_log".
.[E] 2017-10-13T14:37:56Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:37:56Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:37:57Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:37:57Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:37:58Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:37:58Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:37:59Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:37:59Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:00Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:00Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:01Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:01Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:02Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:02Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:03Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:03Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:04Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:04Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
.[E] 2017-10-13T14:38:05Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
[E] 2017-10-13T14:38:05Z keeper.go:526: cannot get configured pg parameters error=pq: the database system is starting up
 done
server started
[I] 2017-10-13T14:38:06Z keeper.go:1403: postgres parameters changed, reloading postgres instance
[I] 2017-10-13T14:38:06Z postgresql.go:249: reloading database configuration
[E] 2017-10-13T14:38:07Z keeper.go:526: cannot get configured pg parameters error=context deadline exceeded
server signaled
[I] 2017-10-13T14:38:08Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[I] 2017-10-13T14:38:08Z keeper.go:1288: already standby
[I] 2017-10-13T14:38:08Z keeper.go:1410: postgres parameters not changed
[I] 2017-10-13T14:38:09Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[I] 2017-10-13T14:38:09Z keeper.go:1288: already standby
[I] 2017-10-13T14:38:09Z keeper.go:1410: postgres parameters not changed
[I] 2017-10-13T14:38:10Z keeper.go:1269: our db requested role is standby followedDB=e8a0366d
[I] 2017-10-13T14:38:10Z keeper.go:1288: already standby
^C

[root@node-1:~]# systemctl start stolon-keeper.service
```

Or in short form, on the standbys I did:

```
systemctl stop stolon-keeper.service
mv /var/lib/postgres/node_1/postgres/recovery.done /var/lib/postgres/node_1/postgres/recovery.conf 
systemctl start stolon-keeper.service
```

**Comment by sgotti** [comment]
@nh2 It's difficult to say what happened with a reproducer or at least also the sentinels logs and the clysterdata (the store contents).

But looks like some db have been promoted to master (recovery.done). This will happen only if the sentinel asked this to the keeper, instead in your case the requested state is standby while to keeper is mastered. Are you using persistent volumes? Are you sure nothing external to the cluster issued a promote on the standbys?

About the bootuid is an uid that changes at every keeper start so the keeper knows when the sentinel has read it's state (to handle changed IP addresses and other races)

**Comment by nh2** [comment]
> It's difficult to say what happened with a reproducer or at least also the sentinels logs and the clysterdata (the store contents).

1. Does `stolonctl clusterdata` simply dump what's stored unter the appropriate key in etcd/consul? If yes, I could just get that out of consul right now.
2. It is unlikely that I can provide a reproduction, because it took weeks for this issue to occur.
3. I think I can provide the keeper+sentinel+proxy logs for the last couple months, (but I wouldn't be able to point out where exactly you'd have to look because I don't know the exact time when the promotion happened that brought things out of order), would this be useful?

> Are you using persistent volumes?

No, I'm running this on plain old dedicated hardware without Kubernetes or so.

> Are you sure nothing external to the cluster issued a promote on the standbys?

Yes, this is a plain Stolon setup, the only thing I do with it is to have my application use Postgres through it, and occasionally reboot some servers. I never change postgres configuration outside of Stolon on it.

**Comment by timfpark** [comment]
I am seeing this as well with a new cluster I just created in Kubernetes.  In my case it happened for the last keeper in the StatefulSet.

**Comment by sgotti** [comment]
@timfpark Can you please open a new issue (since it can be a differenti problem) providing all the keepers and sentinel logs (in different gists), the cluster spec, stolon version and, if reproducible, the steps to reproduce it?

**Comment by danigosa** [comment]
Having same issue as @timfpark, but only happening if I try to get a 5 pods cluster on k8s, happening since last version. Strange thing is that it works well scaling to 3 or 4, it stops scaling at 5 exactly (the last one) with this message:

```
2018-03-28T19:46:17.923Z	WARN	cmd/keeper.go:158	password file permissions are too open. This file should only be readable to the user executing stolon! Continuing...	{"file": "/etc/secrets/stolon/pg_repl_password", "mode": "01000000777"}
2018-03-28T19:46:17.923Z	WARN	cmd/keeper.go:158	password file permissions are too open. This file should only be readable to the user executing stolon! Continuing...	{"file": "/etc/secrets/stolon/pg_su_password", "mode": "01000000777"}
2018-03-28T19:46:17.924Z	INFO	cmd/keeper.go:1914	exclusive lock on data dir taken
2018-03-28T19:46:17.927Z	INFO	cmd/keeper.go:486	keeper uid	{"uid": "keeper4"}
2018-03-28T19:46:17.944Z	INFO	cmd/keeper.go:964	our keeper data is not available, waiting for it to appear
2018-03-28T19:46:22.950Z	INFO	cmd/keeper.go:964	our keeper data is not available, waiting for it to appear
2018-03-28T19:46:27.957Z	INFO	cmd/keeper.go:970	no db assigned
```

`exclusive lock on data dir taken` is something that only appears in the faulty pod, not in the ones successfully started. May this help.

**Comment by sgotti** [comment]
@danigosa I don't see any issue here: maxStandbysPerSender default to 3 (see https://github.com/sorintlab/stolon/blob/master/doc/cluster_spec.md) so tje4th keeper won't have any db assigned.

> exclusive lock on data dir taken is something that only appears in the faulty pod, not in the ones successfully started. May this help.

That's written at keeper startup only.

**Comment by danigosa** [comment]
@sgotti  well then I don't see anything in any log, the thing is that the very same script worked flawlessly for previous version, and for this new one (the current) it works for 3 replicas, 4 replicas it works, but for 5 replicas it does not launch the very last pod with the `no db assigned`. Identical script. Identical docker images. I'm using stolon-chart and custom etcdv3 cluster (from your very own blog post).

**Comment by sgotti** [comment]
@danigosa Sorry but since I don't see any issue here and it's probably not related to this one can you please open a new issue with all the details on the expected behavior and how to reproduce it?

**Comment by danigosa** [comment]
Ok I'll add an issue with my scripts, logs, and etcd dump and steps to it.

**Comment by nh2** [comment]
This is still happening to me now and then on v0.10.0:

```
May 26 03:26:28 node-1 systemd[1]: Started stolon-keeper.service.
May 26 03:26:28 node-1 stolon-keeper[16146]: warning: superuser name and replication user name are the same. Different users are suggested.
May 26 03:26:28 node-1 stolon-keeper[16146]: 2018-05-26T03:26:28.543Z        INFO        cmd/keeper.go:1914        exclusive lock on data dir taken
May 26 03:26:28 node-1 stolon-keeper[16146]: 2018-05-26T03:26:28.543Z        INFO        cmd/keeper.go:486        keeper uid        {"uid": "node_1"}
May 26 03:26:28 node-1 stolon-keeper[16146]: 2018-05-26T03:26:28.547Z        ERROR        cmd/keeper.go:604        cannot get configured pg parameters        {"error": "dial unix /tmp/.s.PGSQL.5432: connect: no such file or directory"}
May 26 03:26:28 node-1 stolon-keeper[16146]: 2018-05-26T03:26:28.548Z        INFO        cmd/keeper.go:978        our db boot UID is different than the cluster data one, waiting for it to be updated        {"bootUUID": "10adfbdd-a96f-4d51-bf25-5c23af6e687d", "clusterBootUUID": "9b31c6f1-c098-4335-8516-b1ced2566b96"}
May 26 03:26:29 node-1 stolon-keeper[16146]: 2018-05-26T03:26:29.547Z        ERROR        cmd/keeper.go:604        cannot get configured pg parameters        {"error": "dial unix /tmp/.s.PGSQL.5432: connect: no such file or directory"}
May 26 03:26:30 node-1 stolon-keeper[16146]: 2018-05-26T03:26:30.548Z        ERROR        cmd/keeper.go:604        cannot get configured pg parameters        {"error": "dial unix /tmp/.s.PGSQL.5432: connect: no such file or directory"}
```

[clusterdata.json.txt](https://github.com/sorintlab/stolon/files/2041253/clusterdata.json.txt)

**Comment by sgotti** [comment]
@nh2 can you provide the sentinels logs? The keeper is waiting for the sentinel to update its bootuid in the cluster data. So for some reasons the sentinel isn't updating it (no leader sentinel, sentinels can't read the keeperinfo from the store etc...)

**Comment by nh2** [comment]
@sgotti OK, here are the sentinel logs around that time:

* [node-1-sentinel-logs.txt](https://github.com/sorintlab/stolon/files/2041615/node-1-sentinel-logs.txt)
* [node-2-sentinel-logs.txt](https://github.com/sorintlab/stolon/files/2041618/node-2-sentinel-logs.txt)
* [node-3-sentinel-logs.txt](https://github.com/sorintlab/stolon/files/2041619/node-3-sentinel-logs.txt)

**Comment by sgotti** [comment]
@nh2 around `May 26 03:26` I see the sentinels having errors talking to consul and failing leader election. So that's the reason the sentinel isn't updating clusterdata. If you have further data let's continue on bitter or the mailing list. In this issue I can't see a real bug and I'd like to close it since it's old.

**Comment by nh2** [comment]
> since it's old

Well it's not really old, it happened again just this night at 5 am bringing me out of bed ;)

> I see the sentinels having errors talking to consul and failing leader election. So that's the reason the sentinel isn't updating clusterdata.

That's right, but consul has long recovered since then and seems to work fine.

My understanding is that if Consul is temporarily down or out of quorum, Stolon shouldn't get stuck (as it is for me now), but resume operation as soon as Consul has recovered.

**Comment by nh2** [comment]
Oh, I might have found what cause the above.

My logs I uploaded in the last post don't actually go far back enough.

Looks like the error appeared first as part of a temporary `no space left on device`. 
[keeper-node-1-out-of-disk-error.txt](https://github.com/sorintlab/stolon/files/2041771/keeper-node-1-out-of-disk-error.txt)

Excerpt:

```
already standby
postgres parameters not changed
postgres hba entries not changed
our db requested role is standby        {"followedDB": "063ec7b0"}
already standby
postgres parameters not changed
postgres hba entries not changed
our db requested role is standby        {"followedDB": "063ec7b0"}
already standby
postgres parameters not changed
postgres hba entries not changed
 failed to save db local state        {"error": "write /var/lib/postgres/node_1/dbstate009560124: no space left on device"}
our db requested role is standby        {"followedDB": "063ec7b0"}
already standby
 failed to start postgres        {"error": "error writing conf file: write /var/lib/postgres/node_1/postgres/postgresql.conf236793038: no space left on device"}
cannot get configured pg parameters        {"error": "dial unix /tmp/.s.PGSQL.5432: connect: no such file or directory"}
cannot get configured pg parameters        {"error": "dial unix /tmp/.s.PGSQL.5432: connect: no such file or directory"}
our db requested role is standby        {"followedDB": "063ec7b0"}
 cannot move from master role to standby role
```

**Comment by nh2** [comment]
For the workaround for the above problem to recover my node-1: @sgotti said on Gitter:

> If you want to recover now you should ensure that the keeper isn't the current master, stop the keeper, remove it from the cluster (stolonctl removerlkeeper), delete the datadir and start it.

Our keeper UID to remove is `node_1`, as seen in `stolonctl --cluster-name stolon-cluster --store-backend=consul status`.

```
[root@node-1:~]# systemctl stop stolon-keeper
[root@node-1:~]# stolonctl --cluster-name stolon-cluster --store-backend=consul removekeeper node_1
[root@node-1:~]# mv /var/lib/postgres/node_1/ /var/lib/postgres/node_1.trash
[root@node-1:~]# systemctl start stolon-keeper
```

Then I confirmed the keeper looked right in `stolonctl --cluster-name stolon-cluster --store-backend=consul status`, and cleaned up with

```
[root@node-1:~]# rm -r /var/lib/postgres/node_1.trash
```

**Comment by sgotti** [comment]
@nh2 PR #495

**Comment by sgotti** [comment]
To summarize this issue: in case of fs full the recovery.conf while wasn't correctly atomically written ending with an empty recovery.conf file, so the keeper considered this node as a primary and refused to continue since the cluster data says that is should be a standby.

Fixed in #495
